data_model/derive/src/lib.rs

```diff
@@ -14,216 +96,578 @@ use quote::{format_ident, quote};
+///                    | (Self::ByMetadataRemoved, crate::prelude::DomainEvent::MetadataRemoved(_)) => {
+///                        true
+///                    }
+///                (Self::ByAccount(filter_opt), crate::prelude::DomainEvent::Account(event)) => {
+///                    filter_opt.matches(event)
+///                }
+///                (
+///                    Self::ByAssetDefinition(filter_opt),
+///                    crate::prelude::DomainEvent::AssetDefinition(event),
+///                ) => filter_opt.matches(event),
+///                _ => false,
+///            }
+///        }
+///    }
+/// ```
 #[proc_macro_derive(Filter)]
 pub fn filter_derive(input: TokenStream) -> TokenStream {
-    let ast = syn::parse(input).expect("Failed to parse input TokenStream.");
-    impl_from_filter(&ast)
+    let event = parse_macro_input!(input as filter::EventEnum);
+    filter::impl_filter(&event)
 }
 
-fn impl_from_filter(ast: &syn::DeriveInput) -> TokenStream {
-    // Omitting attributes as they don't need to be inherited for filters
-    let syn::DeriveInput {
-        vis,
-        ident: event_ident,
-        generics,
-        data,
-        ..
-    } = ast;
-
-    let event_filter_ident = format_ident!("{}Filter", event_ident);
-
-    let filter_ident = format_ident!(
-        "{}Filter",
-        event_ident
-            .to_string()
-            .strip_suffix("Event")
-            .expect("Events should follow the naming format")
-    );
-
-    let event_variants = if let syn::Data::Enum(syn::DataEnum { variants, .. }) = data {
-        variants.iter().collect::<Vec<_>>()
-    } else {
-        panic!("Only `...Event` enums are supported")
-    };
-
-    let mut filter_variants_idents_id_fields = Vec::<syn::Ident>::new();
-    let mut filter_variants_idents_event_fields = Vec::<syn::Ident>::new();
-
-    let import_path = quote! { crate::prelude };
-
-    let filter_variants_event_fields = event_variants
-        .iter()
-        .filter_map(|variant| {
-            let event_filter_variant_ident = format_ident!("By{}", variant.ident);
-            if let syn::Fields::Unnamed(ref unnamed) = variant.fields {
-                let variant_type = &unnamed
-                    .unnamed
-                    .first()
-                    .expect("Should have at least one type")
-                    .ty;
-                process_event_variant(
-                    variant_type,
-                    &event_filter_variant_ident,
-                    &mut filter_variants_idents_id_fields,
-                    &mut filter_variants_idents_event_fields,
-                )
-                .map(|inner_filter_ident| {
-                    Some(quote! {
-                        #event_filter_variant_ident (#import_path::FilterOpt<#inner_filter_ident>)
-                    })
+mod id {
+    use super::*;
+
+    pub(super) struct IdInput {
+        _attrs: Vec<Attribute>,
+        _vis: Visibility,
```

Comment:
why do you have `_` prefix?

Reply:
Even though the `attrs` were used for finding the attribute with `parse_id_attribute`, they weren't used in the structure per se. Removed that from the struct entirely.

Reply:
And what about the other variables? I see this pattern is pervasive

Reply:
Done, just left it out of the `Parse`d structs